distributed/comm/inproc.py

```diff
@@ -0,0 +1,335 @@
+                   )
+
+
+logger = logging.getLogger(__name__)
+
+ConnectionRequest = namedtuple('ConnectionRequest',
+                               ('c2s_q', 's2c_q', 'c_loop', 'c_addr',
+                                'conn_event'))
+
+
+class Manager(object):
+    """
+    An object coordinating listeners and their addresses.
+    """
+
+    def __init__(self):
+        self.listeners = weakref.WeakValueDictionary()
+        self.addr_suffixes = itertools.count(1)
+        self.ip = get_ip()
+        self.lock = threading.Lock()
+
+    def add_listener(self, addr, listener):
+        with self.lock:
+            if addr in self.listeners:
+                raise RuntimeError("already listening on %r" % (addr,))
+            self.listeners[addr] = listener
+
+    def remove_listener(self, addr):
+        with self.lock:
+            del self.listeners[addr]
+
+    def get_listener_for(self, addr):
+        with self.lock:
+            self.validate_address(addr)
+            return self.listeners.get(addr)
+
+    def new_address(self):
+        return "%s/%d/%s" % (self.ip, os.getpid(), next(self.addr_suffixes))
+
+    def validate_address(self, addr):
+        """
+        Validate the address' IP and pid.
+        """
+        ip, pid, suffix = addr.split('/')
+        if ip != self.ip or int(pid) != os.getpid():
+            raise ValueError("inproc address %r does not match host (%r) or pid (%r)"
+                             % (addr, self.ip, os.getpid()))
+
+
+global_manager = Manager()
+
+def new_address():
+    """
+    Generate a new address.
+    """
+    return 'inproc://' + global_manager.new_address()
+
+
+class QueueEmpty(Exception):
+    pass
+
+
+class Queue(object):
+    """
+    A single-reader, single-writer, non-threadsafe, peekable queue.
+    """
+
+    def __init__(self):
+        self._q = deque()
+        self._read_future = None
+
+    def get_nowait(self):
+        q = self._q
+        if not q:
+            raise QueueEmpty
+        return q.popleft()
+
+    def get(self):
+        assert not self._read_future, "Only one reader allowed"
+        fut = Future()
+        q = self._q
+        if q:
+            fut.set_result(q.popleft())
+        else:
+            self._read_future = fut
+        return fut
+
+    def put_nowait(self, value):
+        q = self._q
+        fut = self._read_future
+        if fut is not None:
+            assert len(q) == 0
+            self._read_future = None
+            fut.set_result(value)
+        else:
+            q.append(value)
+
+    put = put_nowait
+
+    _omitted = object()
+
+    def peek(self, default=_omitted):
+        """
+        Get the next object in the queue without removing it from the queue.
+        """
+        q = self._q
+        if q:
+            return q[0]
+        elif default is not self._omitted:
+            return default
+        else:
+            raise QueueEmpty
+
+
+def _maybe_deserialize(msg):
```

Comment:
I would rather avoid such code in the communication layer. @mrocklin, do you think it's ok to put this in `distributed.protocol`?